core/src/dht/dht_inner_reducers.rs

```diff
@@ -0,0 +1,135 @@
+    let new_status_eav = create_crud_status_eav(old_address, CrudStatus::Modified)?;
+    (*store.meta_storage().write()?).add_eavi(&new_status_eav)?;
+    // add link from old to new
+    let crud_link_eav = create_crud_link_eav(old_address, new_address)?;
+    (*store.meta_storage().write()?).add_eavi(&crud_link_eav)?;
+
+    Ok(new_address.clone())
+}
+
+pub(crate) fn reduce_remove_entry_inner(
+    store: &mut DhtStore,
+    latest_deleted_address: &Address,
+    deletion_address: &Address,
+) -> HcResult<Address> {
+    // pre-condition: Must already have entry in local content_storage
+    let content_storage = &store.content_storage().clone();
+
+    let maybe_json_entry = content_storage.read()?.fetch(latest_deleted_address)?;
+    let json_entry = maybe_json_entry.ok_or_else(|| {
+        HolochainError::ErrorGeneric(String::from("trying to remove a missing entry"))
+    })?;
+
+    let entry = Entry::try_from(json_entry).expect("Stored content should be a valid entry.");
+    // pre-condition: entry_type must not by sys type, since they cannot be deleted
+    if entry.entry_type().to_owned().is_sys() {
+        return Err(HolochainError::ErrorGeneric(String::from(
+            "trying to remove a system entry type",
+        )));
+    }
+    // pre-condition: Current status must be Live
+    // get current status
+    let meta_storage = &store.meta_storage().clone();
+    let status_eavs = meta_storage.read()?.fetch_eavi(&EaviQuery::new(
+        Some(latest_deleted_address.clone()).into(),
+        Some(Attribute::CrudStatus).into(),
+        None.into(),
+        IndexFilter::LatestByAttribute,
+    ))?;
+
+    //TODO clean up some of the early returns in this
+    // TODO waiting for update/remove_eav() assert!(status_eavs.len() <= 1);
+    // For now checks if crud-status other than Live are present
+    let status_eavs = status_eavs
+        .into_iter()
+        .filter(|e| CrudStatus::from_str(String::from(e.value()).as_ref()) != Ok(CrudStatus::Live))
+        .collect::<BTreeSet<EntityAttributeValueIndex>>();
+    if !status_eavs.is_empty() {
+        return Err(HolochainError::ErrorGeneric(String::from(
+            "entry_status != CrudStatus::Live",
+        )));
+    }
+    // Update crud-status
+    let result = create_crud_status_eav(latest_deleted_address, CrudStatus::Deleted);
+    if result.is_err() {
+        return Err(HolochainError::ErrorGeneric(String::from(
+            "Could not create eav",
+        )));
+    }
+    let new_status_eav = result.expect("should unwrap eav");
+    let meta_storage = &store.meta_storage().clone();
+
+    (*meta_storage.write()?).add_eavi(&new_status_eav)?;
+
+    // Update crud-link
+    let crud_link_eav = create_crud_link_eav(latest_deleted_address, deletion_address)
+        .map_err(|_| HolochainError::ErrorGeneric(String::from("Could not create eav")))?;
+    (*meta_storage.write()?).add_eavi(&crud_link_eav)?;
+
+    Ok(latest_deleted_address.clone())
+}
```

Comment:
Awesome, I think because it is now an inner function we could utilize the '?' and get rid of some to those early 'return error' blocks.

Reply:
I was able to get rid of some the early returns using `.map_err` but I left the ones where they return early based on a boolean condition (e.g. 
```
if entry.entry_type().to_owned().is_sys() {
        return Err(HolochainError::ErrorGeneric(
            "trying to remove a system entry type".into()
        ));
 }
``` 

I couldn't figure out a more readable way to write this. Any ideas?

Reply:
imo I think this is fine to leave in! It was nice to clean out some of those other returns tho

This looks great